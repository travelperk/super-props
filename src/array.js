import { error } from "./utils";
import { randomNumber } from "./utils";

export default function array(
  shape,
  {
    nullable = false,
    allowEmpty = true,
    minLength = 0,
    maxLength = Infinity,
    unique = false
  } = {}
) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (!Array.isArray(value))
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not an array.`
      );

    if (value.length === 0 && !allowEmpty)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be an empty array.`
      );

    if (value.length < minLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be shorter than ${minLength}.`
      );

    if (value.length > maxLength)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` can't be longer than ${maxLength}.`
      );

    value.forEach((item, index) => {
      shape({ [index]: item }, index, `${componentName}[]`);
    });

    if (unique && value.length !== Array.from(new Set(value)).length)
      throw error(
        `Property \`${propName}\` of \`${componentName}\` has repeated values.`
      );
  }

  type.toString = function() {
    return `array(${shape})`;
  };

  type.make = function() {
    const length = randomNumber({
      min: minLength,
      max: maxLength === Infinity ? 10 : maxLength,
      integer: true
    });

    const generatedArray = [];
    for (let i = 0; i < length; i++) {
      let element;
      do {
        element = shape.make();
      } while (unique && isElementInArray(element, generatedArray));
      generatedArray.push(element);
    }
    return generatedArray;
  };

  return type;
}

function isElementInArray(element, array) {
  return array.some(e => e === element);
}
