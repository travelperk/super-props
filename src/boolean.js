import { error, choose } from "./utils";

export default function boolean({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "boolean")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a boolean.`
      );
  }

  type.toString = function() {
    return "boolean()";
  };

  type.make = function make() {
    return choose([true, false]);
  };

  return type;
}
