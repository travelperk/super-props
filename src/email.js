import { string } from ".";

export default function email(options) {
  const pattern = /^[a-z0-9!#$%&'*+/=?^_`{|}~.-]+@[a-z0-9-]+(\.[a-z0-9-]+)*$/i;

  const type = string({ ...options, pattern });

  type.toString = function() {
    return "string()";
  };

  return type;
}
