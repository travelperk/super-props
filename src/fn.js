import { error } from "./utils";

export default function fn({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value == null && nullable) return;

    if (typeof value !== "function")
      throw error(
        `Property \`${propName}\` of \`${componentName}\` is not a function.`
      );
  }

  type.toString = function() {
    return "fn()";
  };

  return type;
}
