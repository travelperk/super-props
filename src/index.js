export { default as boolean } from "./boolean";
export { default as number } from "./number";
export { default as integer } from "./integer";
export { default as string } from "./string";
export { default as email } from "./email";
export { default as object } from "./object";
export { default as literal } from "./literal";
export { default as oneOf } from "./oneOf";
export { default as fn } from "./fn";
export { default as array } from "./array";
export { default as thrower } from "./thrower";
