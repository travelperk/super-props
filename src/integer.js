import number from "./number";

export default function integer(opts = {}) {
  const type = number({ integer: true, ...opts });

  type.toString = function() {
    return "integer()";
  };

  return type;
}
