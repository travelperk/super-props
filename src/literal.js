import { boolean } from ".";
import { error } from "./utils";

export default function value(literalValue) {
  if (typeof literalValue === "object" && literalValue !== null)
    throw new Error("literal value can't be an object");

  if (typeof literalValue === "function")
    throw new Error("literal value can't be a function");

  function type(props, propName, componentName) {
    const value = props[propName];

    if (value !== literalValue)
      throw error(
        `Property \`${propName}.\` of \`${componentName}\` was supposed to be \`${literalValue}\` but was \`${value}\`.`
      );
  }

  type.make = function() {
    return literalValue;
  };

  type.toString = function() {
    return `literal(${literalValue})`;
  };

  return type;
}
