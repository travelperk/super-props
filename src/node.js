import PropTypes from "prop-types";
import { error } from "./utils";

export default function node({ nullable = false } = {}) {
  function type(props, propName, componentName) {
    const value = props[propName];
    if (value === null) return;

    const validator = nullable ? PropTypes.node : PropTypes.node.isRequired;

    const validationError = validator(
      props,
      propName,
      componentName,
      "",
      "",
      "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED"
    );

    if (validationError) throw new error(validationError.message);
  }

  type.toString = function() {
    return "node()";
  };

  return type;
}
