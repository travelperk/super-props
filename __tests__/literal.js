import { literal } from "../src";

describe("literal()", () => {
  const values = [true, false, null, undefined, "foo", 1, 1.2];

  describe("make()", () => {
    it.each(values)("should return the passed value %p", value => {
      expect(literal(value).make()).toBe(value);
    });
  });

  it.each(values)("should validate the passed value %p", value => {
    const valueType = literal(value);
    values.forEach(testValue => {
      if (testValue === value)
        expect(() =>
          valueType({ testValue }, "testValue", "Component")
        ).not.toThrow();
      else
        expect(() =>
          valueType({ testValue }, "testValue", "Component")
        ).toThrowErrorMatchingSnapshot();
    });
  });

  it("should throw if it recieves an object", () => {
    expect(() => literal({})).toThrow();
    expect(() => literal({ foo: "bar" })).toThrow();
  });

  it("should throw if it recieves a function", () => {
    expect(() => literal(() => {})).toThrow();
  });
});
