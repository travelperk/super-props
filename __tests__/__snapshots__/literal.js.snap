// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`literal() should validate the passed value "foo" 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`true\`."`;

exports[`literal() should validate the passed value "foo" 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`false\`."`;

exports[`literal() should validate the passed value "foo" 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`null\`."`;

exports[`literal() should validate the passed value "foo" 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`undefined\`."`;

exports[`literal() should validate the passed value "foo" 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`1\`."`;

exports[`literal() should validate the passed value "foo" 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`foo\` but was \`1.2\`."`;

exports[`literal() should validate the passed value 1 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`true\`."`;

exports[`literal() should validate the passed value 1 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`false\`."`;

exports[`literal() should validate the passed value 1 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`null\`."`;

exports[`literal() should validate the passed value 1 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`undefined\`."`;

exports[`literal() should validate the passed value 1 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`foo\`."`;

exports[`literal() should validate the passed value 1 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1\` but was \`1.2\`."`;

exports[`literal() should validate the passed value 1.2 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`true\`."`;

exports[`literal() should validate the passed value 1.2 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`false\`."`;

exports[`literal() should validate the passed value 1.2 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`null\`."`;

exports[`literal() should validate the passed value 1.2 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`undefined\`."`;

exports[`literal() should validate the passed value 1.2 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`foo\`."`;

exports[`literal() should validate the passed value 1.2 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`1.2\` but was \`1\`."`;

exports[`literal() should validate the passed value false 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`true\`."`;

exports[`literal() should validate the passed value false 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`null\`."`;

exports[`literal() should validate the passed value false 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`undefined\`."`;

exports[`literal() should validate the passed value false 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`foo\`."`;

exports[`literal() should validate the passed value false 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`1\`."`;

exports[`literal() should validate the passed value false 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`false\` but was \`1.2\`."`;

exports[`literal() should validate the passed value null 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`true\`."`;

exports[`literal() should validate the passed value null 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`false\`."`;

exports[`literal() should validate the passed value null 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`undefined\`."`;

exports[`literal() should validate the passed value null 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`foo\`."`;

exports[`literal() should validate the passed value null 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`1\`."`;

exports[`literal() should validate the passed value null 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`null\` but was \`1.2\`."`;

exports[`literal() should validate the passed value true 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`false\`."`;

exports[`literal() should validate the passed value true 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`null\`."`;

exports[`literal() should validate the passed value true 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`undefined\`."`;

exports[`literal() should validate the passed value true 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`foo\`."`;

exports[`literal() should validate the passed value true 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`1\`."`;

exports[`literal() should validate the passed value true 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`true\` but was \`1.2\`."`;

exports[`literal() should validate the passed value undefined 1`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`true\`."`;

exports[`literal() should validate the passed value undefined 2`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`false\`."`;

exports[`literal() should validate the passed value undefined 3`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`null\`."`;

exports[`literal() should validate the passed value undefined 4`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`foo\`."`;

exports[`literal() should validate the passed value undefined 5`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`1\`."`;

exports[`literal() should validate the passed value undefined 6`] = `"[TKType] Property \`testValue.\` of \`Component\` was supposed to be \`undefined\` but was \`1.2\`."`;
